File: haskell/sublist/src/Sublist.hs
module Sublist (Sublist(..), sublist) where
import Data.List (isInfixOf)

-- The task is to create the data type `Sublist`, with `Eq` and
-- `Show` instances, and implement the function `sublist`.

data Sublist = Sublist | Superlist | Equal | Unequal deriving (Eq, Show)

sublist :: (Eq a) => [a] -> [a] -> Sublist
sublist xs ys
    | xs == ys          = Equal
    | xs `isInfixOf` ys = Sublist
    | ys `isInfixOf` xs = Superlist
    | otherwise         = Unequal

File: clojure/sublist/src/sublist.clj
(ns sublist)


(defn sublist? [xs ys]
  (->> ys
    (partition (count xs) 1)
    (some #{xs})))


(defn classify [list1 list2]
  (cond (= list1 list2)         :equal
        (sublist? list1 list2)  :sublist
        (sublist? list2 list1)  :superlist
        :else                   :unequal))
